### PR TITLE
Developer swatelogic

### PR DESCRIPTION
**Pull request by Freymaurer**
Will come with some utility functions:

- `import { Templates } from 'arctrl'` (please verify import from index.js works correctly), has functions to filter arrays of templates.
  - `Templates.getDistinctTags()`, `Templates.getDistinctEndpointRepositories()`, `getDistinctOntologyAnnotations()`
  - `Templates.filterByTags(tagArray)`, also exists for EndpointRepositories and OntologyAnnotations. Can be used with optional boolean input to toggle between OR and AND search. Default is false, is OR.

```js
const templates = [create_TestTemplate()];
const queryTags = [OntologyAnnotation.fromString("DNA"), OntologyAnnotation.fromString("RNA")];
Templates.filterByOntologyAnnotation(queryTags, true)(templates); //example for AND search
```

- `CompositeHeader` and `IOType` both have `getUITooltip(caseName)`, which will return a string with a description. `caseName` can be used from `CompositeHeader.Cases`/`IOType.Cases`
- `Templates_fromJsonString ` will now return an array.
- `ArcTable.Join`, can be used to append one ArcTable to another. See source code comment. ⚠️for js `TableJoinOptions` can be any of "Headers" (Add only headers, no values), "WithUnit" (Add headers and unit information without main value), "WithValues" (Add full columns).

```fsharp
/// <summary>
/// This function can be used to join two arc tables.
/// </summary>
/// <param name="table">The table to join to this table.</param>
/// <param name="joinOptions">Can add only headers, header with unitized cell information, headers with values.</param>
/// <param name="forceReplace">if set to true will replace unique columns.</param>
member this.Join(table:ArcTable, ?joinOptions: TableJoinOptions, ?forceReplace: bool) : unit =
```

### CODE REVIEW

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 1]
```diff
@@ -426,8 +435,49 @@ type ArcTable(name: string, headers: ResizeArray<CompositeHeader>, values: Syste
...
 
-    //member this.InsertParameterValue () =
-    //    this.in
+    /// <summary>
+    /// This function can be used to join two arc tables.
+    /// </summary>
+    /// <param name="table">The table to join to this table.</param>
+    /// <param name="joinOptions">Can add only headers, header with unitized cell information, headers with values.</param>
+    /// <param name="forceReplace">if set to true will replace unique columns.</param>
+    member this.Join(table:ArcTable, ?joinOptions: TableJoinOptions, ?forceReplace: bool) : unit =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:444]: I like!

File: src/ARCtrl/Templates/Templates.fs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        oas
+        |> Array.ofSeq
+        |> Array.distinct
+
+    /// <summary>
+    /// Filter templates by `template.Tags`.
+    /// </summary>
+    /// <param name="queryTags">The ontology annotation to filter by.</param>
+    /// <param name="isAnd">Default: false. If true all `queryTags` must be contained in template, if false only 1 tags must be contained in template.</param>
+    static member filterByTags(queryTags: OntologyAnnotation [], ?isAnd: bool) =
```
[the remark this answers is not in the document]
Comment (HLWeil) [reply src/ARCtrl/Templates/Templates.fs:59]: Only here. In the context of `getDefaultComparer` it becomes more obvious what it's for.

**Review by HLWeil — CHANGES_REQUESTED** [review]

File: tests/ISA/ISA.Tests/CompositeHeader.Tests.fs [hunk 3]
```diff
@@ -26,11 +26,19 @@ let private tests_iotype =
...
-            let count = caseInfos |> Array.length
-            Expect.hasLength IOType.All (count-1) "Expect one less than all because we do not want to track `FreeText` case."
+        testCase "Cases" (fun () -> 
+            let caseInfos = IOType.Cases
+            Expect.hasLength IOType.All (caseInfos.Length-1) "Expect one less than all because we do not want to track `FreeText` case."
         )
+        testCase "getExplanation" <| fun _ ->
+            let cases = IOType.Cases |> Array.map snd
+            for case in cases do
+                let actual = IOType.getExplanation(U2.Case2 case)
```
Comment (HLWeil) [inline tests/ISA/ISA.Tests/CompositeHeader.Tests.fs:36]: Seems a bit weird in F#. Any other workaround?

File: src/ARCtrl/Templates/Templates.fs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                match isValid, contains with
+                | None, any -> isValid <- Some any
+                | Some maybe, any -> isValid <- Some (comparer maybe any)
+            Option.defaultValue false isValid
+        )
+
+[<AttachMembers>]
+type Templates =
+
+    static member getDistinctTags (templates: Template []) =
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/ARCtrl/Templates/Templates.fs:31]: ups

File: src/ISA/ISA/ArcTypes/ArcTypes.fs [hunk 5]
```diff
@@ -49,6 +49,10 @@ module ArcTypesAux =
             | None ->
                 ()
 
+    let boxHashOption (a: 'a option) : obj =
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/ArcTypes.fs:52]: Does not work anyways, i cannot get the custom GetHashCode() to work correctly 😞

File: src/ISA/ISA/ArcTypes/ArcTypes.fs [hunk 6]
```diff
@@ -1616,6 +1649,18 @@ type ArcInvestigation(identifier : string, ?title : string, ?description : strin
             copy.DeregisterAssay(studyIdentifier, assayIdentifier)
             copy
 
+    /// <summary>
+    /// Returns all Contacts/Performers from assays/studies/investigation unfiltered. 
+    /// </summary>
+    member this.GetAllPersons() =
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/ArcTypes.fs:1655]: I also thought about this, but how do you decide a distinct? full distinct? disctinct by orcid? The API consumer can easily decide any disctinct they want to apply. In my opinion best to leave it like this.

File: src/ISA/ISA/ArcTypes/CompositeColumn.fs [hunk 7]
```diff
@@ -38,3 +38,14 @@ type CompositeColumn = {
             // Maybe still return `msg` somehow if `raiseExeption` is false?
             false
 
+    /// <summary>
+    /// Returns an array of all units found in the cells of this column. Returns None if no units are found.
+    /// </summary>
+    member this.GetColumnUnits() =
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/CompositeColumn.fs:44]: true

File: tests/ISA/ISA.Tests/ArcAssay.Tests.fs [hunk 8]
```diff
@@ -598,6 +598,31 @@ let private test_UpdateBy = testList "UpdateBy" [
         Expect.equal actual.Performers [|yield! expected.Performers; yield! next.Performers|] "Performers"
         Expect.equal actual.Comments [|yield! expected.Comments; yield! next.Comments|] "Comments"
 ]
+
+let private tests_GetHashCode = testList "GetHashCode" [
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply tests/ISA/ISA.Tests/ArcAssay.Tests.fs:602]: just does not work. Neither on dotnet or fable. I have no idea how to solve this. 

```fsharp
override this.GetHashCode() = failwith "Still need to implement correct GetHashCode function for ArcInvestigation type."
    //[|
    //    box this.Identifier
    //    ArcTypesAux.boxHashOption this.Title
    //    ArcTypesAux.boxHashOption this.Description
    //    ArcTypesAux.boxHashOption this.SubmissionDate
    //    ArcTypesAux.boxHashOption this.PublicReleaseDate
    //    this.Publications
    //    this.Contacts
    //    this.OntologySourceReferences
    //    this.Assays
    //    this.Studies
    //    this.RegisteredStudyIdentifiers
    //    this.Comments
    //    this.Remarks
    //|]
    //|> Array.map (fun o -> o.GetHashCode())
    //|> Array.sum
```

This is how GetHashCode for assay/study/investigation looks like.

File: tests/ISA/ISA.Tests/CompositeHeader.Tests.fs (hunk 3 above)
Comment (Freymaurer) [reply tests/ISA/ISA.Tests/CompositeHeader.Tests.fs:36]: `getExplanation`/`GetExplanation` is a pure UI feature. Its design aims to prove a iterative method for showing all cases and respective explanation. 

If you dislike the design, we can move it to a more hidden `UI` module without flooding our main API space.

Comment (HLWeil) [reply tests/ISA/ISA.Tests/CompositeHeader.Tests.fs:36]: If it's really exclusively meant for the UI, how about adjusting the naming to accomodate this? E.g. "getUIToolTip" or smt like this

File: src/ISA/ISA/ArcTypes/ArcTypes.fs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/ArcTypes.fs:52]: Moved!

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 9]
```diff
@@ -712,7 +713,19 @@ type ArcTable(name: string, headers: ResizeArray<CompositeHeader>, values: Syste
 
     // it's good practice to ensure that this behaves using the same fields as Equals does:
     override this.GetHashCode() = 
-        let name = this.Name.GetHashCode()
-        let headers = this.Headers |> Seq.fold (fun state ele -> state + ele.GetHashCode()) 0
-        let bodyCells = this.Values |> Seq.sortBy (fun x -> x.Key) |> Seq.fold (fun state ele -> state + ele.GetHashCode()) 0
-        name + headers + bodyCells
\ No newline at end of file
+        //let v1,v2 = 
+        let v =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:717]: hashcode does ignore order?
If so, does equality behave the same way?

**Review by HLWeil — CHANGES_REQUESTED** [review]
I would like one more test for ArcTable and ArcAssay:

Inequality using `Equals`() when also inequality using `GetHashCode()`

File: src/ISA/ISA/ArcTypes/ArcTable.fs (hunk 9 above)
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/ArcTable.fs:717]: i have a very nice test for table equality, but unsure how do good tests for inequality on assay,study,investigation. Have a look

**Review by HLWeil — APPROVED** [review]

**Pull request merged**